src/heredoc/heredoc.c

```diff
@@ -95,69 +25,110 @@ char	*get_hd_file_name(void)
+ * 
+ * 2. SIGUSR1: Sets up the signal handler for SIGUSR1 to handle_sigint_heredoc
+ * Thanks to that when the child process sends it to indicate that the heredoc
+ * process has been interrupted by SIGINT (Ctrl + C), the parent process saves
+ * this information in the global variable g_signal. This variable is then
+ * checked in the minishell_loop before the execution is triggered and if
+ * it is set to SIGUSR1, the execution won't be triggered because SIGINT
+ * received during the heredoc process cancels the whole command.
+ * 
+ * 
+ * @param	current_redirect		list containing heredoc content
+ * @param	content_simle_cmd		pointer to the content of current_simple_cmd
+ * 									to access the name of the temp heredoc file
+ * @param	data					pointer to the t_data structure
+ * 									(for exit_status and to send to 
+ * 									create_heredoc and then expander)
  */
-void	process_heredoc(t_list *simple_cmd, t_data *data)
+void	process_heredoc(t_data *data, t_list *current_redirect,
+	t_simple_cmds *content_simple_cmd)
 {
-	t_simple_cmds	*content_simple_cmd;
-	t_list			*current_redirect;
-	t_lexer			*content_redirect;
+	int				pid;
+	int				status;
 
-	content_simple_cmd = (t_simple_cmds *) simple_cmd->content;
-	current_redirect = content_simple_cmd->redirects;
-	while (current_redirect)
+	pid = fork();
+	if (pid == -1)
 	{
-		content_redirect = (t_lexer *) current_redirect->content;
-		if (content_redirect->token == LESS_2)
-		{
-			if (content_simple_cmd->hd_file)
-				free(content_simple_cmd->hd_file);
-			content_simple_cmd->hd_file = get_hd_file_name();
-			create_heredoc(current_redirect, content_simple_cmd->hd_file, data);
-		}
-		current_redirect = current_redirect->next;
+		ft_putendl_fd("minishell: fork: Resource temporarily unavailable", 2);
+		exit_current_prompt(NULL);
+	}
+	if (pid == 0)
+	{
+		create_heredoc(current_redirect, content_simple_cmd->hd_file,
+			data);
+		exit_minishell(NULL, 0);
+	}
+	else
+	{
+		signal(SIGINT, SIG_IGN);
+		signal(SIGUSR1, handle_sigint_heredoc);
+		waitpid(pid, &status, 0);
+		if (WIFEXITED(status))
+			data->exit_status = WEXITSTATUS(status);
 	}
 }
 
 /**
- * @brief   Executes the heredoc process for the entire command sequence.
+ * @brief   Searches for heredoc content in the whole simple_cmds sequence and
+ * executes the heredoc process for each command.
+ *
+ * The function traverses the simple commands list and for each one of that it
+ * traverses the redirects list via a nested while loop. 
+ * 
+ * For each heredoc redirection, <<, the process_heredoc function is called.
  * 
- * The function traverses the simple commands list, sending each node into
- * the process_heredoc which takes care of running the whole heredoc process if
- * a heredoc redirection, <<, is encountered.
+ * When << is encountered and the hd_file already exists, it is freed as to
+ * make space for the new (just now encountered) heredoc because only the last
+ * heredoc should be taken into account.
```

Comment:
The main thing that changed in heredoc is that the child process now takes care of the writting into the file. The rest of the changes is just moving things around, mostly because of logic and norminette.